### DISCUSSION TITLE
Usar Yarn ou NPM?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [frontendbr/forum#1790] [Perguntas e Respostas]
Galera, 
além e culturas e costumes, há de fato vantagens e desvantagens entre NPM e Yarn? Quais vocês usam? O que vocês veem de diferenças?

**Comment by felipe-heredia** [comment]
Eu não sei se realmente existe vantagem (maquinaria) entre um ou outro.

Mas utilizo o Yarn, vejo ele de uma forma mais bonita, não gostei muito de como o NPM me retorna no console, preferi o Yarn mesmo.

Esse é o motivo pelo qual utilizo o Yarn, mesmo que comprovado que o NPM é mais rápido, pensaria bastante antes de utilizar.

**Comment by waldemirflj** [comment]
@wilsonneto-dev tem mais haver com preferência e familiaridade. Eu utilizo o NPM, já utiliza ele antes do Yarn... hehehe

**Comment by rodrigodiasnoronha** [comment]
Ah, ai vai da sua preferência. Eu gosto do Yarn, mas gosto ainda mais do NPM.

**Comment by ninetails** [comment]
Ambos são bons, ambos tem vantagens e desvantagens (se comparar as versões atuais). Mas com a chegada do [berry](https://github.com/yarnpkg/berry) comecei a gostar menos do Yarn.

**TL;DR** Yarn pra monorepos (e/ou Yarn Plug'n'Play, pq o Tink vai ter um plug'n'play mas será diferente... ainda não fui atrás), npm pro restante.

### Por que Yarn pra _monorepos_?

[Workspaces](https://classic.yarnpkg.com/en/docs/workspaces/). Ele unifica dependência de todos os pacotes do _monorepo_ em um só _node_modules_ (ou _cache folder_ no PnP), inclusive fazendo _dedupe_. Claro, precisa ativar o _Yarn Workspaces_ pra funcionar na v1.x, mas ainda sim tem muita coisa que se repete entre pacotes: lint & seus configs, babel & configs, jest etc...

---

Tirando isso, só se quiser algo bem mais específico do Yarn, tipo [Offline Mirror](https://classic.yarnpkg.com/en/docs/offline-mirror/). Ou [npm audit](https://docs.npmjs.com/cli/audit) no caso do npm.

**Reply by tcelestino** [reply]
Com a implementação do npm workspace, ainda é vantagem utilizar o Yarn para monorepo?

**Reply by ninetails** [reply]
Tentei usar o npm workspaces quando saiu... algumas libs (como o Lerna) ainda não davam suporte, pois a implementação ainda é diferente da do Yarn.

Se tiverem corrigido esses problemas, não vejo mais Yarn como motivo pra usar pra monorepo... Ou até diria que se quem estiver dando manutenção no teu monorepo, que faça com que funcione nos dois (por Developer Experience).

**Reply by vitebo** [reply]
eu to querendo atualizar um repositório lerna aqui do npm 6 para o 7, posso trazer mais informações depois caso aconteça algum problema. Até então eu acredito que não vai ter problemas porque o próprio repo do lerna atualmente ta usando os workspaces do npm 7.

**Reply by tcelestino** [reply]
> Tentei usar o npm workspaces quando saiu... algumas libs (como o Lerna) ainda não davam suporte, pois a implementação ainda é diferente da do Yarn.
> 
> Se tiverem corrigido esses problemas, não vejo mais Yarn como motivo pra usar pra monorepo... Ou até diria que se quem estiver dando manutenção no teu monorepo, que faça com que funcione nos dois (por Developer Experience).

@ninetails estou estruturando uma aplicação que já existe, porém temos uma dependência que não tá fazendo sentindo ficar fora dessa aplicação. Como a aplicação já tem o npm, não estou vendo vantagens de trocar pelo Yarn. Olhando alguns projetos que o próprio Lerna lista no seu site, alguns já estão usando o npm.

---

> eu to querendo atualizar um repositório lerna aqui do npm 6 para o 7, posso trazer mais informações depois caso aconteça algum problema. Até então eu acredito que não vai ter problemas porque o próprio repo do lerna atualmente ta usando os workspaces do npm 7.

Boa @vitebo, vai mandando o feedback aqui na thread. 🚀

**Reply by ninetails** [reply]
opa, tacam-lhe pau então :D

Eu ultimamente não ando mais codando muito com frontend, digo mesmo que só testei o npm workspaces quando saiu.

Sei que a comunidade que contribui pro Lerna é bem próxima dos dois package managers, então creio que devam ter corrigido sim.

**Reply by vitebo** [reply]
Só pra deixar registrado aqui, tivemos zero problemas em utilizar o lerna com o npm@7.
se ajudar alguem no futuro a implementar tbm, o repositório alvo foi esse aqui: https://github.com/quero-edu/guidelines
mas não tem nada de muito diferente não
